src/storage/mvcc/txn.rs

```diff
@@ -1311,26 +1317,125 @@ mod tests {
+        // Currently we cannot avoid this.
+        must_acquire_pessimistic_lock(&engine, k, k, 26, 29);
+        must_pessimistic_rollback(&engine, k, 26, 29);
+        must_unlocked(&engine, k);
+
+        // Non pessimistic key in pessimistic transaction.
+        must_pessimistic_prewrite_put(&engine, k, v, k, 30, 30, false);
+        must_locked(&engine, k, 30);
+        must_commit(&engine, k, 30, 31);
+        must_unlocked(&engine, k);
+        must_get_commit_ts(&engine, k, 30, 31);
+
+        // Rollback collapsed.
+        must_rollback_collapsed(&engine, k, 32);
+        must_rollback_collapsed(&engine, k, 33);
+        must_acquire_pessimistic_lock_err(&engine, k, k, 32, 32);
+        // Currently we cannot avoid this.
+        must_acquire_pessimistic_lock(&engine, k, k, 32, 34);
+        must_pessimistic_rollback(&engine, k, 32, 34);
+        must_unlocked(&engine, k);
+
+        // Acquire lock when there is lock with different for_update_ts.
+        must_acquire_pessimistic_lock(&engine, k, k, 35, 36);
+        must_pessimistic_locked(&engine, k, 35, 36);
+        must_acquire_pessimistic_lock(&engine, k, k, 35, 35);
+        must_pessimistic_locked(&engine, k, 35, 36);
+        must_acquire_pessimistic_lock(&engine, k, k, 35, 37);
+        must_pessimistic_locked(&engine, k, 35, 37);
+
+        // Cannot prewrite when there is another transaction's pessimistic lock.
+        must_pessimistic_prewrite_put_err(&engine, k, v, k, 36, 36, true);
+        must_pessimistic_prewrite_put_err(&engine, k, v, k, 36, 38, true);
+        must_pessimistic_locked(&engine, k, 35, 37);
+        // Cannot prewrite when there is another transaction's non-pessimistic lock.
+        must_pessimistic_prewrite_put(&engine, k, v, k, 35, 37, true);
+        must_locked(&engine, k, 35);
+        must_pessimistic_prewrite_put_err(&engine, k, v, k, 36, 38, true);
+        must_locked(&engine, k, 35);
+
+        // Commit pessimistic transaction's key but with smaller commit_ts than for_update_ts.
+        // Currently not checked, so in this case it will actually be successfully committed.
+        must_commit(&engine, k, 35, 36);
+        must_unlocked(&engine, k);
+        must_get_commit_ts(&engine, k, 35, 36);
+
+        // Prewrite meets pessimistic lock on a non-pessimistic key.
+        // Currently not checked, so prewrite will success.
+        must_acquire_pessimistic_lock(&engine, k, k, 40, 40);
+        must_pessimistic_locked(&engine, k, 40, 40);
+        must_pessimistic_prewrite_put(&engine, k, v, k, 40, 40, false);
+        must_locked(&engine, k, 40);
+        must_commit(&engine, k, 40, 41);
+        must_unlocked(&engine, k);
+
+        // Prewrite with different for_update_ts.
+        // Currently not checked.
+        must_acquire_pessimistic_lock(&engine, k, k, 42, 45);
+        must_pessimistic_locked(&engine, k, 42, 45);
+        must_pessimistic_prewrite_put(&engine, k, v, k, 42, 43, true);
```

Comment:
Normally, the `for_update_ts` in `pessimistic_perwrite` is greater than or equal to lock's because TiDB uses the last `for_update_ts` in the transaction to prewrite.